src/Compilers/CSharp/Portable/Binder/RefSafetyAnalysis.cs

```diff
@@ -594,6 +594,63 @@ static uint getDeclarationValEscape(BoundTypeExpression typeExpression, uint val
         public override BoundNode? VisitRecursivePattern(BoundRecursivePattern node)
         {
             SetPatternLocalScopes(node);
+
+            if (node.DeconstructMethod is { } deconstructMethod)
+            {
+                using (new PatternInput(this, _localScopeDepth))
+                {
+                    base.VisitRecursivePattern(node);
+                }
+
+                var placeholders = ArrayBuilder<(BoundValuePlaceholderBase, uint)>.GetInstance();
+
+                var receiver = new BoundDeconstructValuePlaceholder(
+                    node.Syntax,
+                    variableSymbol: null,
+                    isDiscardExpression: false,
+                    node.InputType);
+                placeholders.Add((receiver, _localScopeDepth));
+
+                ImmutableArray<BoundExpression> arguments = node.Deconstruction.SelectAsArray(static (x, placeholders) =>
+                {
+                    if (x.Pattern is BoundObjectPattern { VariableAccess: { } variableAccess })
+                    {
+                        return variableAccess;
+                    }
+
+                    var placeholder = new BoundDeconstructValuePlaceholder(
+                        x.Syntax,
+                        variableSymbol: null,
+                        isDiscardExpression: true,
```

Comment:
I didn't understand why we should pass 'isDiscardExpression: true' here.

Reply:
Because this placeholder corresponds to an `out _` discard parameter in the lowered `Deconstruct` call.

Reply:
Originally this code path was written to handle discards in the pattern which correspond to `out _` discard arguments of the equivalent `Deconstruct` call.

But now I realize there can be also nested subpatterns and we need to be smarter about extracting all variable accesses from that.